File: connection-examples/go/state.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
)

type SingleValue struct {
	Value interface{} `json:"value"`
}

type Update struct {
	value interface{} `json:"value"`
	diff  int64 `json:"diff"`
}

type State struct {
	state     map[string]int64
	timestamp int64
	valid     bool
	history   []Update
}

func NewState(collectHistory bool) *State {
	state := make(map[string]int64)
	history := []Update{}

	return &State{
		state:     state,
		timestamp: 0,
		valid:     true,
		history:   history,
	}
}

func (s *State) getState() []interface{} {
	list := []interface{}{}

	for key, value := range s.state {
		clone := make(map[string]interface{})
		err := json.Unmarshal([]byte(key), &clone)
		if err != nil {
			fmt.Println(err);
			continue
		}

		for i := int64(0); i < value; i++ {
			list = append(list, clone["value"])
		}
	}

	return list
}

func (s *State) getHistory() []Update {
	return s.history
}

func (s *State) validate(timestamp int64) error {
	if !s.valid {
		return errors.New("Invalid state.")
	} else if timestamp < s.timestamp {
		s.valid = false
		return errors.New("Invalid timestamp.")
	}
	return nil
}

func (s *State) process(update Update) {
	var sv = SingleValue { Value: update.value }
	value, err := json.Marshal(sv)
	// fmt.Println(sv, value);
	if err != nil {
		fmt.Println(err);
		return
	}

	count, ok := s.state[string(value)]
	if !ok {
		count = 0
	}

	count += update.diff

	if count <= 0 {
		delete(s.state, string(value))
	} else {
		s.state[string(value)] = count
	}

	if s.history != nil {
		s.history = append(s.history, update)
	}
}

func (s *State) Update(updates []Update, timestamp int64) error {
	if len(updates) > 0 {
		err := s.validate(timestamp)
		if err != nil {
			return err
		}
		s.timestamp = timestamp
		for _, update := range updates {
			s.process(update)
		}
	}
	return nil
}

File: connection-examples/python/pg8000/state.py
from typing import TypeVar, List, Union
import json

T = TypeVar('T')

class Update:
    def __init__(self, value: T, diff: int):
        self.value = value
        self.diff = diff

class State:
    def __init__(self, collect_history: bool = False):
        self.state = {}
        self.timestamp = 0
        self.valid = True
        self.history = [] if collect_history else None

    def get_state(self) -> List[T]:
        result = []
        for key, value in self.state.items():
            clone = json.loads(key)
            for i in range(int(value)):
                result.append(clone)
        return result

    def get_history(self):
        return self.history

    def validate(self, timestamp: int):
        if not self.valid:
            raise Exception("Invalid state.")
        elif timestamp < self.timestamp:
            print("Invalid timestamp.")
            self.valid = False
            raise Exception(f"Update with timestamp ({timestamp}) is lower than the last timestamp ({self.timestamp}). Invalid state.")

    def process(self, update: Update):
        value = json.dumps(update['value'])
        count = self.state.get(value, 0) + update['diff']

        if count <= 0:
            del self.state[value]
        else:
            self.state[value] = count

        if self.history is not None:
            self.history.append(update)

    def update(self, updates: List[Update], timestamp: int):
        if len(updates) > 0:
            self.validate(timestamp)
            self.timestamp = timestamp
            for update in updates:
                self.process(update)
